reposerver/repository/repository.go

```diff
@@ -362,31 +364,89 @@ func (s *Service) GenerateManifest(ctx context.Context, q *apiclient.ManifestReq
 		return ok, err
 	}
 
+	tarConcluded := false
+	var promise *ManifestResponsePromise
+
 	operation := func(repoRoot, commitSHA, cacheKey string, ctxSrc operationContextSrc) error {
-		res, err = s.runManifestGen(ctx, repoRoot, commitSHA, cacheKey, ctxSrc, q)
-		return err
+		promise = s.runManifestGen(ctx, repoRoot, commitSHA, cacheKey, ctxSrc, q)
+		select {
+		case err := <-promise.errCh:
+			return err
+		case resp := <-promise.responseCh:
+			res = resp
+		case tarDone := <-promise.tarDoneCh:
+			tarConcluded = tarDone
+		}
+		return nil
 	}
 
 	settings := operationSettings{sem: s.parallelismLimitSemaphore, noCache: q.NoCache, noRevisionCache: q.NoRevisionCache, allowConcurrent: q.ApplicationSource.AllowsConcurrentProcessing()}
-
 	err = s.runRepoOperation(ctx, q.Revision, q.Repo, q.ApplicationSource, q.VerifySignature, cacheFn, operation, settings)
 
+	if tarConcluded && res == nil {
```

Comment:
Can you please add a comment that explains what is going on here :) It took me a while to understand how it works so should summary would be very useful.

Reply:
For sure this is a good place to comment the code to make it more evident what is going on.
Done.